File: cpy0/c/ast_nodes.c
#include "ast_nodes.h"

#include <stdlib.h>

static AstNode *ast_alloc(AstKind kind, int line) {
    AstNode *node = (AstNode *)calloc(1, sizeof(AstNode));
    if (!node) {
        die("out of memory");
    }
    node->kind = kind;
    node->line = line;
    return node;
}

AstNode *ast_new_module(void) {
    AstNode *node = ast_alloc(AST_MODULE, 1);
    ptrvec_init(&node->as.module.body);
    return node;
}

AstNode *ast_new_expr_stmt(AstNode *expr, int line) {
    AstNode *node = ast_alloc(AST_EXPR_STMT, line);
    node->as.expr_stmt.expr = expr;
    return node;
}

AstNode *ast_new_assign(char *name, AstNode *value, int line) {
    AstNode *node = ast_alloc(AST_ASSIGN, line);
    node->as.assign.name = name;
    node->as.assign.value = value;
    return node;
}

AstNode *ast_new_if(AstNode *test, int line) {
    AstNode *node = ast_alloc(AST_IF, line);
    node->as.if_stmt.test = test;
    ptrvec_init(&node->as.if_stmt.body);
    ptrvec_init(&node->as.if_stmt.orelse);
    return node;
}

AstNode *ast_new_while(AstNode *test, int line) {
    AstNode *node = ast_alloc(AST_WHILE, line);
    node->as.while_stmt.test = test;
    ptrvec_init(&node->as.while_stmt.body);
    return node;
}

AstNode *ast_new_function_def(char *name, char **params, int param_count, int line) {
    AstNode *node = ast_alloc(AST_FUNCTION_DEF, line);
    node->as.function_def.name = name;
    node->as.function_def.params = params;
    node->as.function_def.param_count = param_count;
    ptrvec_init(&node->as.function_def.body);
    return node;
}

AstNode *ast_new_return(AstNode *value, int line) {
    AstNode *node = ast_alloc(AST_RETURN, line);
    node->as.return_stmt.value = value;
    return node;
}

AstNode *ast_new_pass(int line) {
    return ast_alloc(AST_PASS, line);
}

AstNode *ast_new_name(char *name, int line) {
    AstNode *node = ast_alloc(AST_NAME, line);
    node->as.name.name = name;
    return node;
}

AstNode *ast_new_int(long value, int line) {
    AstNode *node = ast_alloc(AST_CONSTANT, line);
    node->as.constant.int_value = value;
    return node;
}

AstNode *ast_new_float(double value, int line) {
    AstNode *node = ast_alloc(AST_CONSTANT, line);
    node->as.constant.is_float = 1;
    node->as.constant.float_value = value;
    return node;
}

AstNode *ast_new_string(char *value, int line) {
    AstNode *node = ast_alloc(AST_CONSTANT, line);
    node->as.constant.is_string = 1;
    node->as.constant.str_value = value;
    return node;
}

AstNode *ast_new_binop(OpKind op, AstNode *left, AstNode *right, int line) {
    AstNode *node = ast_alloc(AST_BINOP, line);
    node->as.binop.op = op;
    node->as.binop.left = left;
    node->as.binop.right = right;
    return node;
}

AstNode *ast_new_unaryop(OpKind op, AstNode *operand, int line) {
    AstNode *node = ast_alloc(AST_UNARYOP, line);
    node->as.unaryop.op = op;
    node->as.unaryop.operand = operand;
    return node;
}

AstNode *ast_new_compare(OpKind op, AstNode *left, AstNode *right, int line) {
    AstNode *node = ast_alloc(AST_COMPARE, line);
    node->as.compare.op = op;
    node->as.compare.left = left;
    node->as.compare.right = right;
    return node;
}

AstNode *ast_new_call(AstNode *func, int line) {
    AstNode *node = ast_alloc(AST_CALL, line);
    node->as.call.func = func;
    ptrvec_init(&node->as.call.args);
    return node;
}

AstNode *ast_new_attribute(AstNode *value, char *attr, int line) {
    AstNode *node = ast_alloc(AST_ATTRIBUTE, line);
    node->as.attribute.value = value;
    node->as.attribute.attr = attr;
    return node;
}

AstNode *ast_new_subscript(AstNode *value, AstNode *index, int line) {
    AstNode *node = ast_alloc(AST_SUBSCRIPT, line);
    node->as.subscript.value = value;
    node->as.subscript.index = index;
    return node;
}

static void ast_free_vec(PtrVec *vec) {
    for (int i = 0; i < vec->count; i++) {
        ast_free((AstNode *)vec->items[i]);
    }
    ptrvec_free(vec);
}

void ast_free(AstNode *node) {
    if (!node) {
        return;
    }
    switch (node->kind) {
        case AST_MODULE:
            ast_free_vec(&node->as.module.body);
            break;
        case AST_EXPR_STMT:
            ast_free(node->as.expr_stmt.expr);
            break;
        case AST_ASSIGN:
            free(node->as.assign.name);
            ast_free(node->as.assign.value);
            break;
        case AST_IF:
            ast_free(node->as.if_stmt.test);
            ast_free_vec(&node->as.if_stmt.body);
            ast_free_vec(&node->as.if_stmt.orelse);
            break;
        case AST_WHILE:
            ast_free(node->as.while_stmt.test);
            ast_free_vec(&node->as.while_stmt.body);
            break;
        case AST_FUNCTION_DEF:
            free(node->as.function_def.name);
            for (int i = 0; i < node->as.function_def.param_count; i++) {
                free(node->as.function_def.params[i]);
            }
            free(node->as.function_def.params);
            ast_free_vec(&node->as.function_def.body);
            break;
        case AST_RETURN:
            ast_free(node->as.return_stmt.value);
            break;
        case AST_NAME:
            free(node->as.name.name);
            break;
        case AST_CONSTANT:
            if (node->as.constant.is_string) {
                free(node->as.constant.str_value);
            }
            break;
        case AST_BINOP:
            ast_free(node->as.binop.left);
            ast_free(node->as.binop.right);
            break;
        case AST_UNARYOP:
            ast_free(node->as.unaryop.operand);
            break;
        case AST_COMPARE:
            ast_free(node->as.compare.left);
            ast_free(node->as.compare.right);
            break;
        case AST_CALL:
            ast_free(node->as.call.func);
            ast_free_vec(&node->as.call.args);
            break;
        case AST_ATTRIBUTE:
            ast_free(node->as.attribute.value);
            free(node->as.attribute.attr);
            break;
        case AST_SUBSCRIPT:
            ast_free(node->as.subscript.value);
            ast_free(node->as.subscript.index);
            break;
        case AST_PASS:
            break;
    }
    free(node);
}

File: cpy0/rust/src/ast_nodes.rs
use crate::util::PtrVec;

#[derive(Clone, PartialEq, Debug)]
pub enum OpKind {
    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Eq,
    Ne,
    Lt,
    Le,
    Gt,
    Ge,
    Neg,
}

#[derive(Clone, PartialEq, Debug)]
pub enum AstKind {
    Module,
    ExprStmt,
    Assign,
    If,
    While,
    FunctionDef,
    Return,
    Pass,
    Name,
    Constant,
    BinOp,
    UnaryOp,
    Compare,
    Call,
    Attribute,
    Subscript,
}

#[derive(Clone)]
pub struct AstNode {
    pub kind: AstKind,
    pub line: usize,
    pub as_module: Option<ModuleData>,
    pub as_expr_stmt: Option<ExprStmtData>,
    pub as_assign: Option<AssignData>,
    pub as_if_stmt: Option<IfStmtData>,
    pub as_while_stmt: Option<WhileStmtData>,
    pub as_function_def: Option<FunctionDefData>,
    pub as_return_stmt: Option<ReturnStmtData>,
    pub as_name: Option<NameData>,
    pub as_constant: Option<ConstantData>,
    pub as_binop: Option<BinOpData>,
    pub as_unaryop: Option<UnaryOpData>,
    pub as_compare: Option<CompareData>,
    pub as_call: Option<CallData>,
    pub as_attribute: Option<AttributeData>,
    pub as_subscript: Option<SubscriptData>,
}

#[derive(Clone)]
pub struct ModuleData {
    pub body: PtrVec<AstNode>,
}

#[derive(Clone)]
pub struct ExprStmtData {
    pub expr: Box<AstNode>,
}

#[derive(Clone)]
pub struct AssignData {
    pub name: String,
    pub value: Box<AstNode>,
}

#[derive(Clone)]
pub struct IfStmtData {
    pub test: Box<AstNode>,
    pub body: PtrVec<AstNode>,
    pub orelse: PtrVec<AstNode>,
}

#[derive(Clone)]
pub struct WhileStmtData {
    pub test: Box<AstNode>,
    pub body: PtrVec<AstNode>,
}

#[derive(Clone)]
pub struct FunctionDefData {
    pub name: String,
    pub params: Vec<String>,
    pub param_count: usize,
    pub body: PtrVec<AstNode>,
}

#[derive(Clone)]
pub struct ReturnStmtData {
    pub value: Option<Box<AstNode>>,
}

#[derive(Clone)]
pub struct NameData {
    pub name: String,
}

#[derive(Clone)]
pub struct ConstantData {
    pub is_float: bool,
    pub int_value: i64,
    pub float_value: f64,
    pub str_value: Option<String>,
    pub is_string: bool,
}

#[derive(Clone)]
pub struct BinOpData {
    pub op: OpKind,
    pub left: Box<AstNode>,
    pub right: Box<AstNode>,
}

#[derive(Clone)]
pub struct UnaryOpData {
    pub op: OpKind,
    pub operand: Box<AstNode>,
}

#[derive(Clone)]
pub struct CompareData {
    pub op: OpKind,
    pub left: Box<AstNode>,
    pub right: Box<AstNode>,
}

#[derive(Clone)]
pub struct CallData {
    pub func: Box<AstNode>,
    pub args: PtrVec<AstNode>,
}

#[derive(Clone)]
pub struct AttributeData {
    pub value: Box<AstNode>,
    pub attr: String,
}

#[derive(Clone)]
pub struct SubscriptData {
    pub value: Box<AstNode>,
    pub index: Box<AstNode>,
}

impl AstNode {
    pub fn new_module() -> Self {
        AstNode {
            kind: AstKind::Module,
            line: 1,
            as_module: Some(ModuleData {
                body: PtrVec::new(),
            }),
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_expr_stmt(expr: Box<AstNode>, line: usize) -> Self {
        AstNode {
            kind: AstKind::ExprStmt,
            line,
            as_module: None,
            as_expr_stmt: Some(ExprStmtData { expr }),
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_assign(name: String, value: Box<AstNode>, line: usize) -> Self {
        AstNode {
            kind: AstKind::Assign,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: Some(AssignData { name, value }),
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_if(test: Box<AstNode>, line: usize) -> Self {
        AstNode {
            kind: AstKind::If,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: Some(IfStmtData {
                test,
                body: PtrVec::new(),
                orelse: PtrVec::new(),
            }),
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_while(test: Box<AstNode>, line: usize) -> Self {
        AstNode {
            kind: AstKind::While,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: Some(WhileStmtData {
                test,
                body: PtrVec::new(),
            }),
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_function_def(name: String, params: Vec<String>, line: usize) -> Self {
        let param_count = params.len();
        AstNode {
            kind: AstKind::FunctionDef,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: Some(FunctionDefData {
                name,
                params,
                param_count,
                body: PtrVec::new(),
            }),
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_return(value: Option<Box<AstNode>>, line: usize) -> Self {
        AstNode {
            kind: AstKind::Return,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: Some(ReturnStmtData { value }),
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_pass(line: usize) -> Self {
        AstNode {
            kind: AstKind::Pass,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_name(name: String, line: usize) -> Self {
        AstNode {
            kind: AstKind::Name,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: Some(NameData { name }),
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_int(value: i64, line: usize) -> Self {
        AstNode {
            kind: AstKind::Constant,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: Some(ConstantData {
                is_float: false,
                int_value: value,
                float_value: 0.0,
                str_value: None,
                is_string: false,
            }),
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_float(value: f64, line: usize) -> Self {
        AstNode {
            kind: AstKind::Constant,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: Some(ConstantData {
                is_float: true,
                int_value: 0,
                float_value: value,
                str_value: None,
                is_string: false,
            }),
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_string(value: String, line: usize) -> Self {
        AstNode {
            kind: AstKind::Constant,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: Some(ConstantData {
                is_float: false,
                int_value: 0,
                float_value: 0.0,
                str_value: Some(value),
                is_string: true,
            }),
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_binop(op: OpKind, left: Box<AstNode>, right: Box<AstNode>, line: usize) -> Self {
        AstNode {
            kind: AstKind::BinOp,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: Some(BinOpData { op, left, right }),
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_unaryop(op: OpKind, operand: Box<AstNode>, line: usize) -> Self {
        AstNode {
            kind: AstKind::UnaryOp,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: Some(UnaryOpData { op, operand }),
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_compare(op: OpKind, left: Box<AstNode>, right: Box<AstNode>, line: usize) -> Self {
        AstNode {
            kind: AstKind::Compare,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: Some(CompareData { op, left, right }),
            as_call: None,
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_call(func: Box<AstNode>, line: usize) -> Self {
        AstNode {
            kind: AstKind::Call,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: Some(CallData {
                func,
                args: PtrVec::new(),
            }),
            as_attribute: None,
            as_subscript: None,
        }
    }

    pub fn new_attribute(value: Box<AstNode>, attr: String, line: usize) -> Self {
        AstNode {
            kind: AstKind::Attribute,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: Some(AttributeData { value, attr }),
            as_subscript: None,
        }
    }

    pub fn new_subscript(value: Box<AstNode>, index: Box<AstNode>, line: usize) -> Self {
        AstNode {
            kind: AstKind::Subscript,
            line,
            as_module: None,
            as_expr_stmt: None,
            as_assign: None,
            as_if_stmt: None,
            as_while_stmt: None,
            as_function_def: None,
            as_return_stmt: None,
            as_name: None,
            as_constant: None,
            as_binop: None,
            as_unaryop: None,
            as_compare: None,
            as_call: None,
            as_attribute: None,
            as_subscript: Some(SubscriptData { value, index }),
        }
    }
}